### ISSUE TITLE
VelocyPack format

### ISSUE THREAD

**Issue by davechallis** [guoli-lyu/arangors#6]
In case it's of any use/interest, I've started work on a Rust implementation of the VelocyPack format in Rust at: https://github.com/davechallis/rust-velocypack

It's currently got some basic serialization working, currently working on the deserialization part (it's got a long way to go with documentation, decent error messages etc. though of course).

It's implemented using serde, so should hopefully end up not being too painful to work with.

**Comment by fMeow** [comment]
Great! But how can I ask the server to send/receive data in velocy pack?

**Comment by fMeow** [comment]
I find the documentation I need at [velocystream](https://github.com/arangodb/velocystream). That's where to strat off.

**Comment by davechallis** [comment]
Great, glad you found it, it took me a while to find those documents too.

In case it helps, I also used that document to create velocystream data structures in Rust too, here: https://github.com/davechallis/rust-velocystream

It's fairly minimal (this is all the code: https://github.com/davechallis/rust-velocystream/blob/master/src/lib.rs), but might save some time.

**Comment by fMeow** [comment]
Wow, thank you so much!

**Comment by inzanez** [comment]
Just out of interest: I assume that supporting VelocyPack would bring performance boosts? Or is there more about it?

**Comment by fMeow** [comment]
I think VelocyPack can save network bandwidth to some extent.

**Comment by davechallis** [comment]
@inzanez @fMeow yup, it's more compact and faster to serialise, but one of the big advantages is that ArangoDB itself can work with velocypack data internally without conversion, so it's much more efficient from the server side to consume/produce.

There's more detailed description on some of the advantages and design here: https://github.com/arangodb/velocypack#motivation

**Comment by inzanez** [comment]
@davechallis Ok, that sounds quite nice. I will be happy to participate on the VelocyPack implementation if required. I guess we'd need a full implementation so that it could be used.
